src/diff.rs

```diff
@@ -0,0 +1,524 @@
+    let mut rec_plane = input_plane.clone();
+    // Make the test pattern robust to data alignment
+    let xpad_off =
+      (input_plane.cfg.xorigin - input_plane.cfg.xpad) as i32 - 8i32;
+
+    for (i, row) in
+      input_plane.data.chunks_mut(input_plane.cfg.stride).enumerate()
+    {
+      for (j, pixel) in row.into_iter().enumerate() {
+        let val = (j + i) as i32 - xpad_off & 255i32;
+        assert!(
+          val >= u8::min_value().into() && val <= u8::max_value().into()
+        );
+        *pixel = T::cast_from(val);
+      }
+    }
+
+    for (i, row) in rec_plane.data.chunks_mut(rec_plane.cfg.stride).enumerate()
+    {
+      for (j, pixel) in row.into_iter().enumerate() {
+        let val = j as i32 - i as i32 - xpad_off & 255i32;
+        assert!(
+          val >= u8::min_value().into() && val <= u8::max_value().into()
+        );
+        *pixel = T::cast_from(val);
+      }
+    }
+
+    (input_plane, rec_plane)
+  }
+
+  // Regression and validation test for SAD computation
+  fn get_sad_same_inner<T: Pixel>() {
+    let blocks: Vec<(BlockSize, u32)> = vec![
+      (BLOCK_4X4, 1912),
+      (BLOCK_4X8, 4296),
+      (BLOCK_8X4, 3496),
+      (BLOCK_8X8, 7824),
+      (BLOCK_8X16, 16592),
+      (BLOCK_16X8, 14416),
+      (BLOCK_16X16, 31136),
+      (BLOCK_16X32, 60064),
+      (BLOCK_32X16, 59552),
+      (BLOCK_32X32, 120128),
+      (BLOCK_32X64, 186688),
+      (BLOCK_64X32, 250176),
+      (BLOCK_64X64, 438912),
+      (BLOCK_64X128, 654272),
+      (BLOCK_128X64, 1016768),
+      (BLOCK_128X128, 1689792),
+      (BLOCK_4X16, 8680),
+      (BLOCK_16X4, 6664),
+      (BLOCK_8X32, 31056),
+      (BLOCK_32X8, 27600),
+      (BLOCK_16X64, 93344),
+      (BLOCK_64X16, 116384),
+    ];
+
+    let bit_depth: usize = 8;
+    let (input_plane, rec_plane) = setup_planes::<T>();
+
+    for block in blocks {
+      let bsw = block.0.width();
+      let bsh = block.0.height();
+      let area = Area::StartingAt { x: 32, y: 40 };
+
+      let mut input_region = input_plane.region(area);
+      let mut rec_region = rec_plane.region(area);
+
+      assert_eq!(
+        block.1,
+        get_sad(&mut input_region, &mut rec_region, bsw, bsh, bit_depth)
+      );
+    }
+  }
+
+  #[test]
+  fn get_sad_same_u8() {
+    get_sad_same_inner::<u8>();
+  }
+
+  #[test]
+  fn get_sad_same_u16() {
+    get_sad_same_inner::<u16>();
+  }
+
+  fn get_satd_same_inner<T: Pixel>() {
+    let blocks: Vec<(BlockSize, u32)> = vec![
+      (BLOCK_4X4, 1408),
+      (BLOCK_4X8, 2016),
+      (BLOCK_8X4, 1816),
+      (BLOCK_8X8, 3984),
+      (BLOCK_8X16, 5136),
+      (BLOCK_16X8, 4864),
+      (BLOCK_16X16, 9984),
+      (BLOCK_16X32, 13824),
+      (BLOCK_32X16, 13760),
+      (BLOCK_32X32, 27952),
+      (BLOCK_32X64, 37168),
+      (BLOCK_64X32, 45104),
+      (BLOCK_64X64, 84176),
+      (BLOCK_64X128, 127920),
+      (BLOCK_128X64, 173680),
+      (BLOCK_128X128, 321456),
```

Comment:
Also relative to SAD. These look sorta out of whack.

Reply:
Though I guess for a randomly generated frame, it's plausible.

Reply:
This is likely due to how the two planes are set up.